### ISSUE TITLE
A method to support IE11 without having to rely on MutationObserver

### ISSUE THREAD

**Issue by rjgotten** [sdecima/javascript-detect-element-resize#2]
An idea I came up with for my own use:

Create an `<iframe src="about:blank">` and use absolute positioning to stretch it over the full dimensions of whatever element you want to observe for resizes. Clip it with `clip:rect(0,0,0,0)` to become 'invisible' to normal users.

Populate said iframe with an empty HTML document that contains a simple script that defines a listener for the resize event and uses `window.postMessage` to notify the parent window of the resize event.

The message posted by the iframe should encode a known, agreed upon ID from which the parent window can learn which actual element in its document tree had a resize happen. When the parent window receives the message, you can dispatch a local resize event from the element that was resized.

Incase you were wondering: it's possible to synchronously populate the iframe _and_ overcome cross-domain issues by using a simple little trick with the `javascript:` protocol. Set a custom property on the iframe's `contentWindow` that contains whatever markup you want to load into the frame. Then switch the iframe's src like so:

``` javascript
iframe.contentWindow.emitcontent = "<!DOCTYPE html><html><body><script>(...)";
iframe.src = "javascript:window.emitcontent";
```

Please note: to hamper outside tampering, it is best to choose a randomly generated name. The `emitcontent` property is purely for illustration purposes.

This technique has a hidden bonus as well: it will **also** result in a resize event whenever the element switches between visible and invisible status. That may be something of interest to a few people. (Quite a few, I can imagine, infact...)

**Comment by sdecima** [comment]
Thanks for the feedback!
I considered your method and tested it but I finally updated my code using [Back Alley Coder's new implementation using scroll events](http://www.backalleycoder.com/2013/03/18/cross-browser-event-based-element-resize-detection/) to make it work on IE11; I think it is a more standardized solution as I don't have to use a IE11-only fix anymore.

**Comment by rjgotten** [comment]
I used a scroll event solution at first as well (actually took it from GWT; which is where I think BAC's implementation comes from as well), but I quickly dropped it because of its relative brittleness and because of the offsetWidth/-Height readbacks causing superfluous document reflows.

The iframe sensor solution is actually fully cross-browser as well. It's not just an IE11-only fix.

**Comment by sdecima** [comment]
Good point; I didn't see much overhead using the scroll events technique on my tests, but I'll perform further tests with your proposed method and I will reconsider implementing it.

**Comment by bvaughn** [comment]
I'm pretty late to the party. 😄

@rjgotten: I don't suppose you've shared your above solution via some library/util?

I've been using a fork of this library (sdecima/javascript-detect-element-resize) within react-virtualized but the forced reflow is kind of a hard boundary to work around when targeting 60fps.

**Comment by rjgotten** [comment]
@bvaughn 
I actually went back to a scroll event based solution and used DOM read/write batching/queuing to work around the forced reflows.

Iframes turn out to have quite hefty memory costs and while it is possible to swap them out for more lightweight object elements, those have their own compatibility problems in both IE and Safari, especially when used _inside_ an iframe itself.

**Comment by bvaughn** [comment]
Oh nice! Any chance you've shared _that_ approach anywhere I could take a look at and draw inspiration from? :smile:

**Comment by rjgotten** [comment]
@bvaughn 
Sorry. It's in company code, so I cannot publish or share directly.

The principle is fairly simple though;
1. create two queus; one for read operations and one for write operations.
2. Create `read(fn)` and `write(fn)` methods that enqueue functions which will execute read and write operations.
3. When either method is first called during a frame, use `requestAnimationFrame` to schedule a future flush of both queues on the next frame.

(Iirc you're supposed to first flush the write queue and then the read queue to minimize reflows for the amortized case across multiple frames. But I'm not sure.)

**Comment by bvaughn** [comment]
No problem! Just thought I'd ask in case I could save a bit of time. Thanks
for the quick response and the helpful info though!

On Sep 21, 2016 11:23 AM, "rjgotten" notifications@github.com wrote:

@bvaughn https://github.com/bvaughn
Sorry. It's in company code, so I cannot publish or share directly.

The principle is fairly simple though;
1. create two queus; one for read operations and one for write
   operations.
2. Create read(fn) and write(fn) methods that enqueue functions which
   will execute read and write operations.
3. When either method is first called during a frame, use
   requestAnimationFrame to schedule a future flush of both queues on the
   next frame.

(Iirc you're supposed to first flush the write queue and then the read
queue to minimize reflows for the amortized case. But I'm not sure.)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
https://github.com/sdecima/javascript-detect-element-resize/issues/2#issuecomment-248699094,
or mute the thread
